### ISSUE TITLE
Integrate gramaticly correct cues in polish

### ISSUE THREAD

**Issue by paradix** [jonasoreland/runnerup#153]
Polish language has some linguistic specifics when it comes to countable nouns.
We differentiate between nouns for 1, 2-4 and 5 up.

I would like to implement this behaviour for cues in the Formatter class.

**Comment by jonasoreland** [comment]
Good!
See this:
http://developer.android.com/guide/topics/resources/string-resource.html#Plurals

**Comment by jonasoreland** [comment]
please give example

**Comment by paradix** [comment]
Ok, so right now the cues are constructed like this:

activity 8 meters
activity 20 minutes 42 seconds
activity 3 minutes 11 seconds per kilometer
activity 166 bpm
activity heart rate zone 3.4
interval 306 meters
interval 2 minutes 11 seconds
interval 1 minute 12 seconds per kilometer
interval 173 bpm
interval heart rate zone 3.5
lap 690 meters
lap 2 minutes 24 seconds
lap 10 minutes 27 seconds per kilometer
lap 171 bpm
lap heart rate zone 3.4

E.g. in polish it's almost impossible :) to create a sensible-sounding cue with just one word preceding the actual result.
So I was thinking to add the dimension into the cue (like in the HRZ cue), and then the cues would be like this:
activity distance 8 meters
activity time 20 minutes 42 seconds
activity pace 3 minutes 11 seconds per kilometer
activity heart rate 166 bpm
activity heart rate zone 3.4
...

**Comment by jonasoreland** [comment]
that sounds like a good plan.

but i think it should be done in strings.xml somehow,
cause for me (in english) the dimension is just redundant extra noise.

at least I want to cues to be short and distinct, even if they aren't
necessarily grammatically correct.

what do you think ?

/Jonas

**Comment by paradix** [comment]
Well I absolutely get You point of view, but it won't be possible through strings.xml.
Now each cue consists of the Scope.cueId + the value of the dimension. The Scope.cueId is one string value for each enum and in You proposal I would need multiple string values for one cueId depending on which dimension is being processed.

What would You thing about this proposals.

Activity level:
Distance 8 meters
Time 20 minutes 42 seconds
Pace 3 minutes 11 seconds per kilometer
Heart rate 166 bpm
Heart rate zone 3.4

Interval level:
Interval distance 306 meters
Interval time 2 minutes 11 seconds
Interval pace 1 minute 12 seconds per kilometer
Interval heart rate 173 bpm
Interval heart rate zone 3.5

Lap level:
Lap distance 690 meters
Lap time 2 minutes 24 seconds
Lap pace 10 minutes 27 seconds per kilometer
Lap heart rate 171 bpm
Lap heart rate zone 3.4

Or maybe something in this manner would fit.

Activity <-- new cue "header"
Distance 8 meters
Time 20 minutes 42 seconds
Pace 3 minutes 11 seconds per kilometer
Heart rate 166 bpm
Heart rate zone 3.4

Interval <-- new cue "header"
Interval distance 306 meters
Interval time 2 minutes 11 seconds
Interval pace 1 minute 12 seconds per kilometer
Interval heart rate 173 bpm
Interval heart rate zone 3.5

Lap <-- new cue "header"
Lap distance 690 meters
Lap time 2 minutes 24 seconds
Lap pace 10 minutes 27 seconds per kilometer
Lap heart rate 171 bpm
Lap heart rate zone 3.4

**Comment by paradix** [comment]
I absolutely get Your point of view, the cues should be short and distinct.
Doing this through the strings.xml won't be possible, because I need multiple string values for one Scope.cueId (each dimension would require one string).

So maybe something in this manner would fit?

Activity <-- new cue "header"
Distance 8 meters
Time 20 minutes 42 seconds
Pace 3 minutes 11 seconds per kilometer
Heart rate 166 bpm
Heart rate zone 3.4

Interval <-- new cue "header"
Distance 306 meters
Time 2 minutes 11 seconds
Pace 1 minute 12 seconds per kilometer
Heart rate 173 bpm
Heart rate zone 3.5

Lap <-- new cue "header"
Distance 690 meters
Time 2 minutes 24 seconds
Pace 10 minutes 27 seconds per kilometer
Heart rate 171 bpm
Heart rate zone 3.4

**Comment by jonasoreland** [comment]
- not replying to your "full" proposal...sorry...
- having a "scope-header" is definitely something I considered...which
could maybe be good

i'll throw in another "food for thought"-suggestion:
strings.xml could contain "sentences" formated for e.g
http://developer.android.com/reference/java/text/MessageFormat.html

/Jonas

**Comment by paradix** [comment]
I'll definitely will have a look at the MessageFormat but till than I'll push my changes, so You can have a look at the code and the functionality.

**Comment by paradix** [comment]
The MessageFormat proposal works fine, but... it has one flaw concerning the HRZ, which is the only cue which has atm already the dimension included (which makes sense of course).
There might be a problem in formatting this in different languages. It's already problematic in polish :)

The message templates for English would be:
"Activity {0}", "Interval {0}", "Lap {0}"

For Polish it would be something like:
"Activity {1} {0}", "{1} Interval {0}", "{1} lap {0}"

Where:
{0} - value 
{1} - dimension

For HRZ this gives in English: "Interval heart rate zone X.Y"
In Polish (translated to English): "Heart rate zone interval heart rate zone X.Y"

I can imaging this might be also problematic for other languages. Maybe You have an idea for a solution where the value {0} wouldn't have the dimension included?

**Comment by jonasoreland** [comment]
One could make a special case for HRZ...
I.e have a special pattern for it, and an extra if somewhere...

Or...one can make a pattern for each SCOPE/DIMENSION (there is only 3x5 or
so) combination, and override it conditionally per language...

**Comment by paradix** [comment]
It would be nice to know the requirements for other languages to make the right decision. 
I'll think about German as I'm fluent in it and will see what comes up.

**Comment by paradix** [comment]
I think issue is now fixed with leaving English cues unchanged and added the possibility to configure the cues format in the cues.xml

**Comment by paradix** [comment]
Unfortunately I missed the fact that also the Event cues are build with the same cues.xml string that is used for the Dimension cue. So the Scope is basis for both cue types. And with this it's still impossible to construct reasonably sounding cues for both the Event and the Dimension.

Would it be acceptable for You to split the messages for this cues into two different strings?

**Comment by jonasoreland** [comment]
i didn't 100% understand...
but sure, i don't see why not...(atleast not wo/ having seen a patch :)

/Jonas

**Comment by paradix** [comment]
It's not easy to explain but I'll try again.
At this point RunnerUp uses the string from scope.getCueId() in two different use cases.
1) when Event happens (lap starts, activity stops etc.)
2) when any enabled cue is emitted after a period of time or a distance (all the cues from above)

In English it's easy because You always have for: 
1) "Activity started", "Activity paused", "Lap started", etc.
2) "Activity 300 meters", "Lap 4 minutes 32 seconds", etc.
So basically You just need one word per scope - like You already defined it. 

In polish I can't come up with any reasonable solution that could derivative from the English template. That's why I only see a solution in implementing different strings for the cues in 1) and 2).

**Comment by Mixb** [comment]
I propose to consider Idea to implement a correct translation of voice circuits:
1) Create separate string for each pair of activity/interval/lap started/stopped/paused and etc.
In language exists stable combinations of these concepts. It is particularly important for voice circuits, translating single words very badly perceived by sound.
as is: " activity " + " paused ", to be: " activity paused " ;
*A large number of combinations (3х5?),  yet comfortable to be translated into not English languages and less labor-intensive. Implemented using an array.

2) Create special templates for translation of voice circuits. Use special simbols for 
The message templates for English would be:
"Activity {0}", "Interval {0}", "Lap {0}"
For non English: it would be something like:
"Activity {1} {0}", "{1} Interval {0}", "{1} lap {0}"
Where:
{0} - value 
{1} - dimension
*Difficult to understand, hard to to develop. My opinion is 1) variant

**Comment by paradix** [comment]
1) The cue texts are not composed of two separate strings. The tts engine gets the whole sentence, like you proposed. If it sounds badly, it's just the problem of the tts engine.

2) It's not only 3x5 combinations, it's a bit more because this 15 cues you mention are for the start/stop/pause/finish/resume events. You need to consider also the cues you get for your results on interval etc for each dimension. This would be a lot more strings.

But I'm opened for any other code proposals to solve this issue in an acceptable way.

**Comment by gerhardol** [comment]
Is this OK in v1.58?

**Comment by davidedelvento** [comment]
@gerhardol is v1.58 the current beta? If so, please see #607. If not, where is it?

**Comment by gerhardol** [comment]
1.58 is current beta (more or less master), 1.57 is the previous beta, 1.56 the current release on Play

**Comment by davidedelvento** [comment]
Ok so this is for people building from git (which is fine), but probably needs to be documented somewhere.

Note that 1.57 does not exist in github, maybe your or @jonasoreland need to push it from your local repos?

**Comment by gerhardol** [comment]
I believe 1.57 used to be tagged (probably 506d358c40), did Jonas remove the tag?
The beta release was in Play from May to a week ago.

**Comment by davidedelvento** [comment]
Not sure what @jonasoreland did. Neither https://github.com/jonasoreland/runnerup/releases nor https://github.com/jonasoreland/runnerup/tags nor the dropdown menu in https://github.com/jonasoreland/runnerup has v1.57 (they all have v1.56 and v1.58)
